server/src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -1026,13 +1025,9 @@ private static class ClusterStateDiff implements Diff<ClusterState> {
             routingTable = RoutingTable.readDiffFrom(in);
             nodes = DiscoveryNodes.readDiffFrom(in, localNode);
             if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0) && in.readBoolean()) {
-                transportVersions = DiffableUtils.readJdkMapDiff(
-                    in,
-                    DiffableUtils.getStringKeySerializer(),
-                    TRANSPORT_VERSION_VALUE_SERIALIZER
-                );
+                versions = DiffableUtils.readJdkMapDiff(in, DiffableUtils.getStringKeySerializer(), VERSIONS_WRAPPER_VALUE_SERIALIZER);
             } else {
-                transportVersions = null;   // infer at application time
+                versions = null;   // infer at application time
```

Comment:
Should we do this with a non-null `versions` containing a null `transportVersion` field?

Reply:
This is a `Diff<Map<String, VersionsWrapper>>` and I don't know how to accomplish this with diffs. Would this be a diff that, when applied to any `Map<String, VersionsWrapper>>` returns a map where the values are all `VersionsWrappers` with null transport versions?

To back up a little, is there a more standard solution to receiving a diff from an older node that doesn't include a newly added section, and then to fall back to reconstructing the section from other information in the cluster state?

I think I get the general idea of what Diff and Diffable are for. It's for sending small changesets instead of the full many-MB cluster state, right? As far as I can see, diffs for the versions wrapper will mainly be for adding and removing nodes. It looks like what's happening here is that if a diff comes in from a node that doesn't "know" about the transport versions section of cluster state, then we apply that diff by reconstructing this section from discovery nodes. But there won't be a way to do that for system index mappings versions, and probably not for other versions we add in the future.

After a quick search, only `IndexMetadataDiff` seems to deal with a lot of evolution of what the diff can contain. Is there other "prior art"?

Please let me know if I'm overthinking this.

Reply:
Oh sorry I missed that this was in the diff bit of the code. Ok, let's keep it as-is. I'm not sure we really need meaningful diffs here, this data structure is very small, but that's not something to change in this PR.